guacamole/src/main/webapp/app/auth/service/authenticationService.js

```diff
@@ -394,5 +407,45 @@ angular.module('auth').factory('authenticationService', ['$injector',
 
     };
 
+    /**
+     * This function accepts a callback function and stores it in an array of 
+     * functions that will be run before page reload when a user logout is triggered.
+     * 
+     * @param {Function} handlerFunction(resolve, reject)
+     *     The function that we want to run as a callback to the logout event
+     *     The function MUST call either resolve('reason') (success callback) or 
+     *     reject('reason') (failure callback).  This will allow you to do async 
+     *     operations inside the handlerFunction and only trigger success/failure 
+     *     when your operations are finished.
+     *     
+     */
+    service.registerLogoutHandler = function registerLogoutHandler(handlerFunction) {
+        logoutHandlers.push(handlerFunction)
+    }
+
+    /**
+     * This private function transforms all of the registered logout handlers into 
+     * promises which will allow the use of $q.all() or Promise.all() to ensure that 
+     * each handler is activated.
+     */
+    let getLogoutHandlerPromises = function getLogoutHandlerPromises() {
+        let promises = [];
+
+        for (x in logoutHandlers) {
+            let handlerFunction = logoutHandlers[x];
+            promises.push(
+                $q((resolve, reject) => {
+                    try {
+                        handlerFunction(resolve, reject);
+                    } catch (e) {
+                        reject(e);
+                    }
```

Comment:
Won't an object thrown within a promise handler implicitly reject the promise with that object?